src/gluonts/ev/api.py

```diff
@@ -0,0 +1,155 @@
+    Should be replaced by a `ForecastBatch` eventually"""
+
+    def __init__(self, values) -> None:
+        self.values = values
+
+    def __getitem__(self, name):
+        return np.array([self.values[name]])
+
+
+def construct_data(
+    test_data: TestData, predictor: Predictor, **predictor_kwargs
+) -> Iterator[Dict[str, np.ndarray]]:
+    forecasts = predictor.predict(dataset=test_data.input, **predictor_kwargs)
+
+    for input, label, forecast in zip(
+        test_data.input, test_data.label, forecasts
+    ):
+        batching_used = False  # isinstance(forecast, ForecastBatch)
+
+        non_forecast_data = {
+            "label": label["target"],
+            "seasonal_error": seasonal_error(
+                input["target"],
+                seasonality=get_seasonality(freq=forecast.start_date.freqstr),
+            ),
+        }
+        joint_data = ChainMap(non_forecast_data, forecast)
+
+        yield joint_data if batching_used else EvaluationDataBatch(joint_data)
+
+
+class MetricEvaluator:
+    def evaluate(
+        self, test_data: TestData, predictor: Predictor, **predictor_kwargs
+    ) -> np.ndarray:
+        data_batches = construct_data(test_data, predictor, **predictor_kwargs)
+
+        for data in data_batches:
+            self.update(data)
+
+        return self.get()
+
+    def update(self, data: Dict[str, np.ndarray]) -> None:
+        raise NotImplementedError
+
+    def get(self) -> np.ndarray:
+        raise NotImplementedError
+
+
+@dataclass
+class StandardMetricEvaluator(MetricEvaluator):
+    """A "standard metric" consists of a metric function and aggregation
+    strategy."""
+
+    map: Callable
+    aggregate: Aggregation
+
+    def update(self, data: Dict[str, np.ndarray]) -> None:
+        self.aggregate.step(self.map(data))
+
+    def get(self) -> np.ndarray:
+        return self.aggregate.get()
+
+
+@dataclass
+class DerivedMetricEvaluator(MetricEvaluator):
+    """A "derived metric" depends on the prior calculation of "standard
+    metrics"."""
+
+    metrics: Dict[str, StandardMetricEvaluator]
+    post_process: Callable
+
+    def update(self, data: Dict[str, np.ndarray]) -> None:
+        for metric in self.metrics.values():
+            metric.update(data)
+
+    def get(self) -> np.ndarray:
+        return self.post_process(
+            **{name: metric.get() for name, metric in self.metrics.items()}
+        )
+
+
+@dataclass
+class MultiMetricEvaluator:
+    """Allows feeding in data once and calculating multiple metrics"""
+
+    metric_evaluators: Dict[str, MetricEvaluator] = field(default_factory=dict)
+
+    def add_metric(self, metrics: Metric, axis: Optional[int] = None) -> None:
+        self.add_metrics([metrics], axis)
+
+    def add_metrics(
+        self, metrics: Collection[Metric], axis: Optional[int] = None
+    ) -> None:
+        for metric in metrics:
+            metric_evaluator = metric(axis=axis)
+            metric_name = f"{metric.__class__.__name__}[axis={axis}]"
```

Comment:
I think we might want to think about naming a bit more thoroughly.

What about functions, this wouldn't work, would it?

Reply:
Good catch. `__class__.__name__` would just be "function" when metric is a function.

As an easy alternative, the names have to be passed now (see commit [here](https://github.com/awslabs/gluonts/pull/2267/commits/a1c5d994e1787280c6a29e9f1a31e6009e38a3f7)).